File: SemanticEmoji/src/main/java/de/kimanufaktur/nsm/emoji/EmojiGraphBuilder.java
package de.kimanufaktur.nsm.emoji;

import de.kimanufaktur.nsm.decomposition.Concept;
import de.kimanufaktur.nsm.decomposition.graph.conceptCache.GraphUtil;
import org.jgrapht.Graph;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class EmojiGraphBuilder {

    private static GraphBuilder graphBuilder = new GraphBuilder();

    public static Graph getEmojiGraph() throws IOException {
        String emojis = getEmojisAsString();
        Graph graph = graphBuilder.buildGraph(emojis); // use extended GraphBuilder to build special Emoji Graph
        //graphBuilder.saveToGraphML(graph, path);
        return graph;
    }


    public static String getEmojisAsString(){
        /**
         *  reference: http://emojitracker.com/, visited: 20/09/2018, 11:26
         *  #1 face with tears of joy U+1F602
         *  #2 red heart U+2764
         *  #3 recycle U+267B
         *  #4 smiling face with heart eyes U+1F60D
         *  #5 heart symbol emoji, which is used in card games U+2665
         *  #6 sad face with tears streaming down U+1F62D
         *  #7 smiling face with smiling eyes  U+1F60A
         *  #8 face is not amused U+1F612
         *  #9 Two pink love hearts U+1F495
         *  #10 An emoji face blowing a kiss U+1F618
         **/

        return "U+1F62D U+2764 U+267B U+1F60D U+2665 U+1F62D U+1F60A U+1F612 U+1F495 U+1F618";
    }


    public static List<Concept> getEmojisAsConcepts(){
        String emojis = getEmojisAsString();
        List<Concept> concepts = new ArrayList<>();

        for(String e: emojis.split(" ")){
            concepts.add(new Concept(e));
        }
        return concepts;
    }
}

File: semantic_decomposition_experiments/semantic_emoji/emoji_graph_builder.py
from __future__ import annotations
from typing import Dict, List, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from semantic_decomposition import Concept
    from semantic_decomposition.graph.semantic_net import SemanticNet

# Top 10 most frequent emoji codes (Unicode code points)
TOP_EMOJIS: List[str] = [
    "U+1F602",  # 😂 Face with Tears of Joy
    "U+2764",   # ❤️ Red Heart
    "U+1F60D",  # 😍 Smiling Face with Heart-Eyes
    "U+1F629",  # 😩 Weary Face
    "U+1F62D",  # 😭 Loudly Crying Face
    "U+1F618",  # 😘 Face Blowing a Kiss
    "U+1F4AF",  # 💯 Hundred Points
    "U+1F60A",  # 😊 Smiling Face with Smiling Eyes
    "U+1F495",  # 💕 Two Hearts
    "U+1F612",  # 😒 Unamused Face
]

EMOJI_NAMES: Dict[str, str] = {
    "U+1F602": "face_with_tears_of_joy",
    "U+2764":  "red_heart",
    "U+1F60D": "smiling_face_heart_eyes",
    "U+1F629": "weary_face",
    "U+1F62D": "loudly_crying_face",
    "U+1F618": "face_blowing_kiss",
    "U+1F4AF": "hundred_points",
    "U+1F60A": "smiling_face_smiling_eyes",
    "U+1F495": "two_hearts",
    "U+1F612": "unamused_face",
}


class EmojiGraphBuilder:
    """Builds a semantic graph from the top-N most frequent emojis."""

    def get_emojis_as_concepts(self) -> List["Concept"]:
        from semantic_decomposition.concept import Concept
        concepts: List[Concept] = []
        for code in TOP_EMOJIS:
            c = Concept()
            c.litheral = EMOJI_NAMES.get(code, code)
            concepts.append(c)
        return concepts

    def get_emoji_graph(self) -> "SemanticNet":
        from semantic_decomposition.graph.semantic_net import SemanticNet
        from semantic_decomposition.graph.graph_util import GraphUtil
        concepts = self.get_emojis_as_concepts()
        if not concepts:
            return SemanticNet()
        graph = GraphUtil.create_graph(concepts[0])
        for concept in concepts[1:]:
            g = GraphUtil.create_graph(concept)
            graph = GraphUtil.merge_graph(graph, g)
        return graph
